Add a page that explains the basics of the Dynamixel communication-system. (#305)
NUbots/NUbook: src/book/02-system/04-hardware/04-dynamixel-protocol.mdx

The goal of this guide is to give an overview of the Dynamixel communication. It is not meant to be an exhaustive list of what is already on the Robotis [manual](https://emanual.robotis.com/docs/en/dxl/protocol2/) but to give further insight into things left out of said manual, mostly the low-level electrical side, as well as any idiosyncracities. Said manual should be read after the reader has read this guide.

The target-audience is somewhat broad and can be anywhere between

- a team-member who is experienced with the Dynamixel system but wants to cross-reference something,
- a new electrical, computer, or mechatronics engineering student who has had some experience with some of these concepts such as UART in ELEC2720 but may want to refresh some of the basics and needs to learn some new things such as RS485 and the Dynamixel protocol,
- and a computer-science student who has some experience with telecommunications such as Ethernet but may have no experience at all with UART and RS485.

<Alert type="warn">

> Assumed knowledge is a basic understanding of electrical concepts like voltage and maybe some HSC physics such as magnetic fields as well as a basic intuition for electronic devices in daily life.

Reviewer: ```suggestion
Assumed knowledge includes a basic understanding of electrical concepts (i.e. voltage), HSC physics (i.e. magnetic fields), and basic intuition for electronic devices in daily life.
```
be definitive and firm in your writing - avoid any 'maybe's. Basic intuition is a bit hard to quantify too. If this is something you think everyone probably has, then maybe just remove that part.

</Alert>

## Universal Asynchronous Receiver-Transmitter

> A universal asynchronous receiver-transmitter (UART) is an electrical device that sends data at an agreed upon bitrate or baudrate. Unlike SPI and I2C where there is one extra signal for the clock, often just called the serial clock, both devices on either end have to know the common baudrate beforehand. Thus, it is common to have to set a baudrate at the very beginning externally either by rewriting firmware or configuring a peripheral.

Reviewer: ```suggestion
A universal asynchronous receiver-transmitter (UART) is an electrical device that sends data at an agreed upon bitrate or baudrate. Unlike [Serial Peripheral Interface (SPI)](https://en.wikipedia.org/wiki/Serial_Peripheral_Interface) and [Inter-Integrated Circuit (I2C)](https://en.wikipedia.org/wiki/I%C2%B2C) where there is one extra signal for the clock, often just called the serial clock, both devices on either end have to know the common baudrate beforehand. It is common with UART to have to set a baudrate at the very beginning externally either by rewriting firmware or configuring a peripheral.
```

<Alert type="info">

> There is also a universal synchronous and asynchronous receiver-transmitter (USART) which is a device that can handle both synchronous and asynchronous communication. Many microcontrollers would have both USARTs and UARTs, the former of which can more of less be used as the latter. However, we will only talk about the asynchronous side in which a UART henceforth can easily be applied for a USART as well.

Reviewer: ```suggestion
There is also a universal synchronous and asynchronous receiver-transmitter (USART) which is a device that can handle both synchronous and asynchronous communication. Many microcontrollers have both USARTs and UARTs, the former of which can more or less be used as the latter. We only consider the asynchronous side in which a UART can easily be applied for a USART as well.
```

</Alert>

> A UART has two pins, an input called RX, and an output called TX.

Reviewer: A UART has two pins: an input called RX and an output called TX.

Reviewer: ```suggestion
A UART has two pins: an input called RX, and an output called TX.
```
unless I'm reading this wrong?

![Two abstract UARTs](./images/dynamixel_protocol/pure_uart.drawio.png 'Two abstract UARTs have their RX-pins and TX-pins connected each way.')

> The term UART can also refer to the way how data is framed. If one will, it can almost be informally thought of as a very low-level and very basic protocol. Here, UART can refer to the special bits that are needed to frame data, all of which need to be agreed upon between both devices on either end.

Reviewer: ```suggestion
The term UART can also refer to the way data is framed. It can be informally thought of as a very low-level and very basic protocol. Here, UART can refer to the special bits that are needed to frame data, all of which need to be agreed upon between both devices on either end.
```

![An 8N1 frame has one start-bit as always, eight bits of data, no parity-bit, and one stop-bit.](./images/dynamixel_protocol/uart_frame.drawio.png 'An 8N1 frame has one start-bit as always, eight bits of data, no parity-bit, and one stop-bit.')

| Line        | Logic-level | Description                                                                 | Optional |
| ----------- | ----------- | --------------------------------------------------------------------------- | -------- |
| Idle        | 1           | is where there is no data.                                                  | no       |
| Start-bit   | 0           | is the beginning of the data.                                               | no       |
| Data-bits   | ?           | are the bits of data from the least significant to the most significant.    | no       |
| Parity-bits | ?           | tell whether there has been any corruption of the data during transmission. | yes      |
| Stop-bit    | 1           | is the end of the data.                                                     | yes      |

> The receiver looks for the start-bit's falling edge. When it has found the edge, it begins to time and to sample each bit thereforth based on the agreed upon bitrate. Most setups are written by three symbols, e.g. 8N1 means that there are eight data-bits, no parity bits, and one stop-bit. Both Dynamixel protocols and many other serial ones are 8N1. Henceforth in this guide, this is the assumed framing.

Reviewer: ```suggestion
The receiver looks for the start-bit's falling edge. When it has found the edge, it begins to time and to sample each bit  based on the agreed upon bitrate. Most setups are described with three symbols, e.g. 8N1 means that there are eight data-bits, no parity bits, and one stop-bit. Both Dynamixel protocols and many other serial ones are 8N1. The rest of this page assumes an 8N1 framing.
```

<Alert type="warn">

Two UARTs must agree on the baudrate, framing (e.g. 8N1), and the signal-levels in order to communicate.

</Alert>

> However, pure UART often only refers to abstract bits or logic levels, not voltages. Signalling techniques such as RS485, RS232, and TTL are what are further needed for a complete system to define the electrical behaviour as well as maybe flow-control (e.g. RS232).

Reviewer: ```suggestion
Pure UART often only refers to abstract bits or logic levels, not voltages. Signalling techniques such as RS485, RS232, and TTL are required for a complete system to define the electrical behaviour as well as optionally flow-control (e.g. RS232).
```

## Duplexity

> Firstly, one must understand the concept of duplexity since the half duplexity of the Dynamixel protocol is not always intuitive for beginners. There are three kinds of digital communication, namely simplex, full duplex, and half duplex.

Reviewer: ```suggestion
There are three kinds of digital communication: simplex, full duplex, and half duplex. Duplexity is important to understand because the Dynamixel protocol uses half duplexity, which is not always intuitive for beginners.
```

Simplexity is when data goes in only one direction, from Device-A to Device-B, where only the former transmits and only the latter receives. There is only one channel, and there may often be only one electrical signal. An analogy would be Alice giving instructions to Bob through a loud-speaker very far away.

![In simplexity, data always flows in only one direction.](./images/dynamixel_protocol/simplexity.drawio.png 'In simplexity, data always flows in only one direction.')

> Full duplexity is when data can go in both directions at the same time, between Device-A and Device-B. There are two channels, one for each direction such that one could think of each channel as simplex. There is no clash when both Device-A and Device-B send data at the same time since the two channels are separate. An analogy would be Alice and Bob writing letters or texting.

Reviewer: ```suggestion
Full duplexity is when data can go in both directions at the same time, between Device-A and Device-B. There are two channels, one for each direction. This is like having two simplex channels. There is no clash when both Device-A and Device-B send data at the same time since the two channels are separate. An analogy would be Alice and Bob writing letters or texting.
```

![In full duplexity, data can flow in both directions at the same time.](./images/dynamixel_protocol/full_duplexity.drawio.png 'In full duplexity, data can flow in both directions at the same time.')

> A pure UART will have two signals, RX and TX, and if it were connected to another UART such that TX->RX both ways, then this would be an example of full duplexity.

Reviewer: ```suggestion
A pure UART will have two signals, RX and TX. RX is used for receiving signals and TX is used for transmitting signals. If two UART devices are connected, they will have two-way communication (TX->RX). This is an example of full duplexity.
```
I tried to make this clearer. My brain just might not be functioning but I looked at this for ages before realising what it was saying.

> Half duplexity is when data can go in both directions but not at the same time. There seem to be two channels but are really combined into one. An analogy would be Alice and Bob talking to each other by walkie-talkies.

Reviewer: ```suggestion
Half duplexity is when data can go in both directions, but not at the same time. There seem to be two channels but they are really combined into one. An analogy would be Alice and Bob talking to each other by walkie-talkies.
```

![In half duplexity, data can flow in both directions but not at the same time.](./images/dynamixel_protocol/half_duplexity.drawio.png 'In half duplexity, data can flow in both directions but not at the same time.')

<Alert type="warn">

The Dynamixel protocol is half duplex; therefore, it is important to understand half duplexity and the needed flow-control and logic behind it.

</Alert>

We will see soon that the reason why the protocol is half duplex is the fewer number of wires - especially for RS485 - and the fact that servos only respond to instructions.

> Often, a Dynamixel device's microcontroller has a UART to communicate on the bus. This UART as said before has two pins, RX and TX. Therefore, in theory, it could be used for full duplexity. However, RX and TX get switched between by a transceiver which is often another chip on the board. This transceiver as the name suggests can switch between receiving and transmitting on the same bus-line. Therefore, this transceiver has another input pin to set the direction, sometimes called the TX-enable, the direction, etc.

Reviewer: ```suggestion
Often, a Dynamixel device's microcontroller has a UART to communicate on the bus. This UART has two pins, RX and TX. Therefore, in theory, it could be used for full duplexity. However, RX and TX get switched between by a transceiver, which is often another chip on the board. The transceiver can switch between receiving and transmitting on the same bus-line. This transceiver has another input pin to set the direction, sometimes called the TX-enable, the direction, etc.
```
I was at first unsure why you said 'often' but then I thought to the case where a dynamixel device is a custom device, like our FSRs. Maybe it'd be useful to mention this.

## Signal-levels

### Transistor-Transistor-Logic

> Transistor-Transistor-Logic (TTL) is a family of logical interfaces mainly for logical circuits. Its name refers to the transistors inside the chip used to control the voltage-level. It also refers to signal-levels for many UARTs and protocols. There are only three wires for a TTL setup for a servo, namely ground, power, and data. The [NC7WZ241](https://www.onsemi.com/pdf/datasheet/nc7wz241-d.pdf) chip is an example of a TTL transceiver that is found the [OpenCR](https://emanual.robotis.com/docs/en/parts/controller/opencr10/).

Reviewer: ```suggestion
Transistor-Transistor-Logic (TTL) is a family of logical interfaces mainly for logical circuits. Its name refers to the transistors inside the chip used to control the voltage-level. It also refers to signal-levels for many UARTs and protocols. There are only three wires for a TTL setup for a servo, namely ground, power, and data. The [NC7WZ241](https://www.onsemi.com/pdf/datasheet/nc7wz241-d.pdf) chip is an example of a TTL transceiver that is found in the [OpenCR](https://emanual.robotis.com/docs/en/parts/controller/opencr10/).
```

![In half duplexity, data can flow in both directions but not at the same time.](./images/dynamixel_protocol/half_duplexity.drawio.png 'In half duplexity, data can flow in both directions but not at the same time.')

<Alert type="warn">

The Dynamixel protocol is half duplex; therefore, it is important to understand half duplexity and the needed flow-control and logic behind it.

</Alert>

The reader will soon see that the reason why the protocol is half duplex is the fewer number of wires - especially for RS485 - and the fact that servos only respond to instructions.

Often, a Dynamixel device's microcontroller has a UART to communicate on the bus. This UART has two pins, RX and TX. Therefore, in theory, it could be used for full duplexity. However, RX and TX get switched between by a transceiver, which is another chip on the board. The transceiver can switch between receiving and transmitting on the same bus-line. This transceiver has another input pin to set the direction, sometimes called the TX-enable or the direction.

## Signal-levels

> However, pure UART often only refers to abstract bits or logic levels, not voltages. Signalling techniques such as RS485, RS232, and TTL are what are further needed for a complete system to define the electrical behaviour as well as the flow-control (e.g. RS232).

Reviewer: Is this missing something before it? Or should the joining "however" be removed?

Author: The latter. I moved that paragraph from another section where there was a paragraph before it. I will fix that tomorrow.
